src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -24,10 +24,298 @@ ucs_status_t uct_tcp_sockcm_ep_disconnect(uct_ep_h ep, unsigned flags)
+    remote_data.dev_addr_length       = addr_length;
+    remote_data.conn_priv_data        = hdr + 1;
+    remote_data.conn_priv_data_length = hdr->length;
+
+    /* the endpoint, passed as the conn_request to the callback, will be passed
+     * to uct_ep_create() which will be invoked by the user and therefore moving
+     * over to its responsibility. */
+    ucs_list_del(&cep->list);
+    cep->listener->conn_request_cb(&cep->listener->super, cep->listener->user_data,
+                                   ifname_str, cep, &remote_data);
+
+    ucs_free(dev_addr);
+    return UCS_OK;
+}
+
+ucs_status_t uct_tcp_sockcm_ep_handle_data_received(uct_tcp_sockcm_ep_t *cep)
+{
+    ucs_status_t status;
+
+    cep->state |= UCT_TCP_SOCKCM_EP_DATA_RECEIVED;
+    uct_tcp_sockcm_ep_init_comm_ctx(cep);
+
+    status = uct_tcp_sockcm_ep_server_invoke_conn_req_cb(cep);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    status = ucs_async_modify_handler(cep->fd, UCS_EVENT_SET_EVWRITE);
+    if (status != UCS_OK) {
+        ucs_error("failed to modify %d event handler to UCS_EVENT_SET_EVWRITE: %s",
+                  cep->fd, ucs_status_string(status));
+        goto out;
+    }
+
+out:
+    return status;
+}
+
+static ucs_status_t uct_tcp_sockcm_ep_recv_nb(uct_tcp_sockcm_ep_t *cep)
+{
+    size_t recv_length;
+    ucs_status_t status;
+
+    recv_length = uct_tcp_sockcm_ep_get_cm(cep)->priv_data_len - cep->comm_ctx.offset;
+    status = ucs_socket_recv_nb(cep->fd, UCS_PTR_BYTE_OFFSET(cep->comm_ctx.buf,
+                                                             cep->comm_ctx.offset),
+                                &recv_length, NULL, NULL);
+    if ((status != UCS_OK) && (status != UCS_ERR_NO_PROGRESS)) {
+        if (status == UCS_ERR_CANCELED) {
+            uct_tcp_sockcm_ep_handle_disconnect(cep, status);
+        } else {
+            ucs_error("ep %p (fd=%d) failed to recv client's data (offset=%zu)",
+                      cep, cep->fd, cep->comm_ctx.offset);
+        }
+        return status;
+    }
+
+    cep->comm_ctx.offset += recv_length;
+    ucs_assertv((cep->comm_ctx.length ?
+                 cep->comm_ctx.offset <= cep->comm_ctx.length : 1), "%zu %zu",
+                cep->comm_ctx.offset, cep->comm_ctx.length);
+    return UCS_OK;
+}
+
+ucs_status_t uct_tcp_sockcm_ep_progress_recv(uct_tcp_sockcm_ep_t *cep)
+{
+    ucs_status_t status;
+
+    status = uct_tcp_sockcm_ep_recv_nb(cep);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    if (uct_tcp_sockcm_ep_is_tx_rx_done(cep)) {
+        status = uct_tcp_sockcm_ep_handle_data_received(cep);
+    }
+
+    return status;
+}
+
+ucs_status_t uct_tcp_sockcm_ep_recv(uct_tcp_sockcm_ep_t *cep)
+{
+    uct_tcp_sockcm_priv_data_hdr_t *hdr;
+    ucs_status_t status;
+
+    status = uct_tcp_sockcm_ep_recv_nb(cep);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    if (cep->comm_ctx.offset < sizeof(*hdr)) {
+        goto out;
+    }
```

Comment:
no need to actually do recv here
we can simplify the code a lot by doing send/recv only from async handler callbacks, would need much less functions calling each other
this function should just initialize the context and prepare to receiving data

Reply:
but uct_tcp_sockcm_ep_recv() is already called only from the async thread...

also, should i remove the sending from the main thread as well?
https://github.com/openucx/ucx/pull/4674/files#diff-0a7c42d102df8d010e7a9bec845edf0dR358

Reply:
yes, can remove both sending and receiving from main thread
main thread should init context and set async handlers from read/write
async thread will wake up and send/recv

Reply:
ok, will remove the sending from the main thread. 
there is no recv from the main frame...